crates/toolchain/platform/custom_insn/src/lib.rs

```diff
@@ -0,0 +1,390 @@
+        }
+        AsmArg::In(tokens) => {
+            template.push_str(", {");
+            template.push_str(reg_name);
+            template.push('}');
+            args.push(quote::quote! { #reg_ident = in(reg) #tokens });
+        }
+        AsmArg::Out(tokens) => {
+            template.push_str(", {");
+            template.push_str(reg_name);
+            template.push('}');
+            args.push(quote::quote! { #reg_ident = out(reg) #tokens });
+        }
+        AsmArg::InOut(tokens) => {
+            template.push_str(", {");
+            template.push_str(reg_name);
+            template.push('}');
+            args.push(quote::quote! { #reg_ident = inout(reg) #tokens });
+        }
+        AsmArg::ConstExpr(tokens) => {
+            template.push_str(", {");
+            template.push_str(reg_name);
+            template.push('}');
+            args.push(quote::quote! { #reg_ident = const #tokens });
+        }
+    }
+}
+
+mod kw {
+    syn::custom_keyword!(In);
+    syn::custom_keyword!(Out);
+    syn::custom_keyword!(InOut);
+    syn::custom_keyword!(Const);
+}
+
+/// Custom RISC-V instruction macro for the zkVM.
+///
+/// This macro is used to define custom R-type RISC-V instructions for the zkVM.
+/// Usage:
+/// ```rust
+/// custom_insn_r!(
+///     opcode = OPCODE,
+///     funct3 = FUNCT3,
+///     funct7 = FUNCT7,
+///     rd = InOut x0,
+///     rs1 = In rs1,
+///     rs2 = In rs2
+/// );
+/// ```
+/// Here, `opcode`, `funct3`, and `funct7` are the opcode, funct3, and funct7 fields of the RISC-V instruction.
+/// `rd`, `rs1`, and `rs2` are the destination register, source register 1, and source register 2 respectively.
+/// The `In`, `Out`, `InOut`, and `Const` keywords are required to specify the type of the register arguments.
+/// They translate to `in(reg)`, `out(reg)`, `inout(reg)`, and `const` respectively, and mean
+/// - "read the value from this variable" before execution (`In`),
+/// - "write the value to this variable" after execution (`Out`),
+/// - "read the value from this variable, then write it back to the same variable" after execution (`InOut`), and
+/// - "use this constant value" (`Const`).
+#[proc_macro]
+pub fn custom_insn_r(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
+    let CustomInsnR {
+        rd,
+        rs1,
+        rs2,
+        opcode,
+        funct3,
+        funct7,
+    } = syn::parse_macro_input!(input as CustomInsnR);
+
+    let mut template = String::from(".insn r {opcode}, {funct3}, {funct7}");
+    let mut args = vec![];
+
+    // Helper function to handle register arguments
+    handle_reg_arg(&mut template, &mut args, &rd, "rd");
+    handle_reg_arg(&mut template, &mut args, &rs1, "rs1");
+    handle_reg_arg(&mut template, &mut args, &rs2, "rs2");
+
+    let expanded = quote::quote! {
+        #[cfg(target_os = "zkvm")]
+        unsafe {
+            core::arch::asm!(
+                #template,
+                opcode = const #opcode,
```

Comment:
I find this weird: why do we do `= const` instead of just generating the whole line as #opcode?